lib/peer.go

```diff
@@ -395,29 +396,52 @@ func (pp *Peer) HandleGetBlocks(msg *MsgDeSoGetBlocks) {
 		return
 	}
 
-	// For each block the Peer has requested, fetch it and queue it to
-	// be sent. It takes some time to fetch the blocks which is why we
-	// do it in a goroutine. This might also block if the Peer's send
-	// queue is full.
-	//
-	// Note that the requester should generally ask for the blocks in the
-	// order they'd like to receive them as we will typically honor this
-	// ordering.
-	//
-	// With HyperSync there is a potential that a node will request blocks that we haven't yet stored, although we're
-	// fully synced. This can happen to archival nodes that haven't yet downloaded all historical blocks. If a GetBlock
-	// is sent to a non-archival node for blocks that we don't have, then the peer is misbehaving and should be disconnected.
-	for _, hashToSend := range msg.HashList {
-		blockToSend := pp.srv.blockchain.GetBlock(hashToSend)
-		if blockToSend == nil {
-			// Don't ask us for blocks before verifying that we have them with a
-			// GetHeaders request.
-			glog.Errorf("Server._handleGetBlocks: Disconnecting peer %v because "+
-				"she asked for a block with hash %v that we don't have", pp, msg.HashList[0])
-			pp.Disconnect()
-			return
+	// Before Version2 we would send each block in a single message, which was quite
+	// slow. Now when we receive a GetBlocks message we will send the blocks in large
+	// batches, which is much faster.
+	if pp.Params.ProtocolVersion == ProtocolVersion2 {
+		allBlocks := MsgDeSoBlockBundle{}
+		for _, hashToSend := range msg.HashList {
+			blockToSend := pp.srv.blockchain.GetBlock(hashToSend)
+			if blockToSend == nil {
+				// Don't ask us for blocks before verifying that we have them with a
+				// GetHeaders request.
+				glog.Errorf("Server._handleGetBlocks: Disconnecting peer %v because "+
+					"she asked for a block with hash %v that we don't have", pp, msg.HashList[0])
+				pp.Disconnect()
+				return
+			}
+			allBlocks.Blocks = append(allBlocks.Blocks, blockToSend)
+		}
+		allBlocks.TipHash = pp.srv.blockchain.blockTip().Hash
+		allBlocks.TipHeight = uint64(pp.srv.blockchain.blockTip().Height)
+		pp.AddDeSoMessage(&allBlocks, false)
+
+	} else {
```

Comment:
Note this logic is exactly what was there before with no changes. Not sure why the diff isn't smart enough to pick that up.